### ISSUE TITLE
关于dropout的疑惑

### ISSUE THREAD

**Issue by bruce5353** [jingyaogong/minimind#614]
作者您好，我注意到config里dropout参数默认为0，当我将其设置成0.3，并对输入执行两次forward的时候，输出却是相同的（我想做一个类似SimCSE，当dropout存在时，对输入forward两次，之后进行对比学习的尝试）。增加的代码是
with autocast ctx:
  res =model(X)
  res2=model(X)
  print(torch.equal(res.last_hidden_state, res2.last_hidden_state))
另外我在model_minimind.py中增加一个基础的测试dropout的模型：
class SimpleModel(nn.Module):
        def __init__(self, p=0.5):
            super().__init__()
            self.dropout = nn.Dropout(p=p)
        
        def forward(self, x):
            return self.dropout(x)
然后增加代码：
with autocast ctx:
  res =model(X)
  res2=model(X)
  print(torch.equal(res.last_hidden_state, res2.last_hidden_state))
  res_simple1 = model_simple(res.last_hidden_state)
  res_simple2 = model_simple(res.last_hidden_state)
  print(torch.equal(res_simple1, res_simple2))

发现res和res2的last_hidden_state是相同的，res_simple1和res_simple2是不同的。这点让我感到很疑惑，希望可以得到大佬您的回答，感谢!

**Comment by jingyaogong** [comment]
`nn.Dropout` 在 `eval()` 模式自动禁用，可以详细了解一下这个问题😊

**Comment by bruce5353** [comment]
作者您好，我还是有点疑惑，因为我是在训练下加的这些代码，请问是哪里启动了eval模式啊

**Comment by jingyaogong** [comment]
> 作者您好，我注意到config里dropout参数默认为0，当我将其设置成0.3，并对输入执行两次forward的时候，输出却是相同的（我想做一个类似SimCSE，当dropout存在时，对输入forward两次，之后进行对比学习的尝试）。增加的代码是 with autocast ctx: res =model(X) res2=model(X) print(torch.equal(res.last_hidden_state, res2.last_hidden_state)) 另外我在model_minimind.py中增加一个基础的测试dropout的模型： class SimpleModel(nn.Module): def **init**(self, p=0.5): super().**init**() self.dropout = nn.Dropout(p=p)
> 
> ```
>     def forward(self, x):
>         return self.dropout(x)
> ```
> 
> 然后增加代码： with autocast ctx: res =model(X) res2=model(X) print(torch.equal(res.last_hidden_state, res2.last_hidden_state)) res_simple1 = model_simple(res.last_hidden_state) res_simple2 = model_simple(res.last_hidden_state) print(torch.equal(res_simple1, res_simple2))
> 
> 发现res和res2的last_hidden_state是相同的，res_simple1和res_simple2是不同的。这点让我感到很疑惑，希望可以得到大佬您的回答，感谢!

---

> 作者您好，我还是有点疑惑，因为我是在训练下加的这些代码，请问是哪里启动了eval模式啊

好吧，第一段描述中 “对输入执行两次forward的时候” 似乎特指eval，我没有get到它在说训练，或者是在哪个阶段。

你可以使用极其简单的方法：
```bash
python3 -c "
import torch
from model.model_minimind import MiniMindConfig, MiniMindForCausalLM

config = MiniMindConfig(hidden_size=512, num_hidden_layers=2, dropout=0.3)
model = MiniMindForCausalLM(config)

X = torch.randint(0, 100, (1, 10))
res1 = model(X)
res2 = model(X)

print(f'training={model.training}, dropout={config.dropout}')
print(f'两次输出相同? {torch.equal(res1.logits, res2.logits)}')
"
```
输出
```
training=True, dropout=0.3
两次输出相同? False
```

---


```bash
python3 -c "
import torch
from model.model_minimind import MiniMindConfig, MiniMindForCausalLM

config = MiniMindConfig(hidden_size=512, num_hidden_layers=2, dropout=0.3)
model = MiniMindForCausalLM(config)
X = torch.randint(0, 100, (1, 10))

# Train 模式
model.train()
res1 = model(X)
res2 = model(X)
print(f'[TRAIN] training={model.training}, 两次输出相同? {torch.equal(res1.logits, res2.logits)}')

# Eval 模式
model.eval()
res3 = model(X)
res4 = model(X)
print(f'[EVAL]  training={model.training}, 两次输出相同? {torch.equal(res3.logits, res4.logits)}')
"
```

输出：
```bash
[TRAIN] training=True, 两次输出相同? False
[EVAL]  training=False, 两次输出相同? True
```

---

从最简单的测试而言，我无法确认你的问题存在，你可以尝试看看，给出更多信息

**Comment by bruce5353** [comment]
感谢作者，我最后排查了一下原因，应该是模型最后输出的self.output出了问题，我将self.output改成output之后就正常了
class MiniMindForCausalLM(PreTrainedModel, GenerationMixin):
    config_class = MiniMindConfig

    def __init__(self, config: MiniMindConfig = None):
        self.config = config or MiniMindConfig()
        super().__init__(self.config)
        self.model = MiniMindModel(self.config)
        self.lm_head = nn.Linear(self.config.hidden_size, self.config.vocab_size, bias=False)
        self.model.embed_tokens.weight = self.lm_head.weight
        self.OUT = CausalLMOutputWithPast()
        self.drout = nn.Dropout(0.3)

    def forward(self,
                input_ids: Optional[torch.Tensor] = None,
                attention_mask: Optional[torch.Tensor] = None,
                past_key_values: Optional[List[Tuple[torch.Tensor, torch.Tensor]]] = None,
                use_cache: bool = False,
                logits_to_keep: Union[int, torch.Tensor] = 0,
                **args):
        h, past_kvs, aux_loss = self.model(
            input_ids=input_ids,
            attention_mask=attention_mask,
            past_key_values=past_key_values,
            use_cache=use_cache,
            **args
        )

        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
        logits = self.lm_head(h[:, slice_indices, :])
        
        # self.OUT.__setitem__('last_hidden_state', h)
        # self.OUT.__setitem__('logits', logits)
        # self.OUT.__setitem__('aux_loss', aux_loss)
        # self.OUT.__setitem__('past_key_values', past_kvs)

        OUT = CausalLMOutputWithPast()
        OUT.__setitem__('last_hidden_state', h)
        OUT.__setitem__('logits', logits)
        OUT.__setitem__('aux_loss', aux_loss)
        OUT.__setitem__('past_key_values', past_kvs)
        return OUT

**Comment by jingyaogong** [comment]
> 感谢作者，我最后排查了一下原因，应该是模型最后输出的self.output出了问题，我将self.output改成output之后就正常了 class MiniMindForCausalLM(PreTrainedModel, GenerationMixin): config_class = MiniMindConfig
> 
> ```
> def __init__(self, config: MiniMindConfig = None):
>     self.config = config or MiniMindConfig()
>     super().__init__(self.config)
>     self.model = MiniMindModel(self.config)
>     self.lm_head = nn.Linear(self.config.hidden_size, self.config.vocab_size, bias=False)
>     self.model.embed_tokens.weight = self.lm_head.weight
>     self.OUT = CausalLMOutputWithPast()
>     self.drout = nn.Dropout(0.3)
> 
> def forward(self,
>             input_ids: Optional[torch.Tensor] = None,
>             attention_mask: Optional[torch.Tensor] = None,
>             past_key_values: Optional[List[Tuple[torch.Tensor, torch.Tensor]]] = None,
>             use_cache: bool = False,
>             logits_to_keep: Union[int, torch.Tensor] = 0,
>             **args):
>     h, past_kvs, aux_loss = self.model(
>         input_ids=input_ids,
>         attention_mask=attention_mask,
>         past_key_values=past_key_values,
>         use_cache=use_cache,
>         **args
>     )
> 
>     slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
>     logits = self.lm_head(h[:, slice_indices, :])
>     
>     # self.OUT.__setitem__('last_hidden_state', h)
>     # self.OUT.__setitem__('logits', logits)
>     # self.OUT.__setitem__('aux_loss', aux_loss)
>     # self.OUT.__setitem__('past_key_values', past_kvs)
> 
>     OUT = CausalLMOutputWithPast()
>     OUT.__setitem__('last_hidden_state', h)
>     OUT.__setitem__('logits', logits)
>     OUT.__setitem__('aux_loss', aux_loss)
>     OUT.__setitem__('past_key_values', past_kvs)
>     return OUT
> ```

然而你贴的是两个月前的旧版本，这个之前已更新

https://github.com/jingyaogong/minimind/blob/9830915d87741b07473265d63c7c74e2286b765e/model/model_minimind.py#L440

**Comment by bruce5353** [comment]
好的，感谢!
